osquery/database/database_plugin.cpp

```diff
@@ -0,0 +1,92 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <osquery/database_plugin.h>
+#include <osquery/logger.h>
+#include <osquery/registry_factory.h>
+#include <osquery/status.h>
+
+#include <osquery/flags.h>
+
+namespace osquery {
+
+Status DatabasePlugin::reset() {
+  // Keep this simple, scope the critical section to the broader methods.
+  tearDown();
+  return setUp();
+}
+
+Status DatabasePlugin::scan(const std::string& domain,
+                            std::vector<std::string>& results,
+                            const std::string& prefix,
+                            size_t max) const {
+  return Status(0, "Not used");
+}
+
+Status DatabasePlugin::call(const PluginRequest& request,
+                            PluginResponse& response) {
+  if (request.count("action") == 0) {
+    return Status(1, "Database plugin must include a request action");
+  }
+
+  // Get a domain/key, which are used for most database plugin actions.
+  auto domain = (request.count("domain") > 0) ? request.at("domain") : "";
+  auto key = (request.count("key") > 0) ? request.at("key") : "";
+
+  if (request.at("action") == "reset") {
+    // Prevent RocksDB reentrancy by logger plugins during plugin setup.
+    VLOG(1) << "Resetting the database plugin: " << getName();
+    auto status = this->reset();
+    if (!status.ok()) {
+      // The active database could not be reset, fallback to an ephemeral.
+      Registry::get().setActive("database", "ephemeral");
+      LOG(WARNING) << "Unable to reset database plugin: " << getName();
+    }
+    return status;
+  }
+
+  // Switch over the possible database plugin actions.
+  if (request.at("action") == "get") {
```

Comment:
There are so many calls `request.at("action")`, would you like to make it once and use it in every condition?

Reply:
(Unrelated to this PR) I really dislike this string-matching-switch API router design. It is so prone to error and the whole project uses it.

NB: it should be possible to keep the thrift API stable and make this routing logic safer throughout the codebase